master/node_selector.go

```diff
@@ -431,48 +444,80 @@ func (s *StrawNodeSelector) getWeight(node Node) float64 {
 	return float64(node.GetAvailableSpace()) / util.GB
 }
 
-func (s *StrawNodeSelector) selectOneNode(nodes []Node) (index int, maxNode Node) {
+func (s *StrawNodeSelector) selectOneNode(nodes []Node, excludedIpSet map[string]struct{}) (index int, maxNode Node) {
 	maxStraw := float64(0)
+	maxStrawNodeIp := ""
 	index = -1
 	for i, node := range nodes {
+		addr := node.GetAddr()
+		ipAndPort := strings.Split(addr, ":")
+		ip := ipAndPort[0]
+
+		if _, ok := excludedIpSet[ip]; ok {
+			continue
+		}
+
 		straw := float64(s.rand.Intn(StrawNodeSelectorRandMax))
 		straw = math.Log(straw/float64(StrawNodeSelectorRandMax)) / s.getWeight(node)
 		if index == -1 || straw > maxStraw {
 			maxStraw = straw
 			maxNode = node
 			index = i
+			maxStrawNodeIp = ip
 		}
 	}
+	if index != -1 {
+		excludedIpSet[maxStrawNodeIp] = struct{}{}
+	}
+
 	return
 }
 
 func (s *StrawNodeSelector) Select(ns *nodeSet, excludeHosts []string, replicaNum int) (newHosts []string, peers []proto.Peer, err error) {
 	nodes := make([]Node, 0)
 	ns.getNodes(s.nodeType).Range(func(key, value interface{}) bool {
 		node := asNodeWrap(value, s.nodeType)
-		if !contains(excludeHosts, node.GetAddr()) {
-			nodes = append(nodes, node)
+		if contains(excludeHosts, node.GetAddr()) {
+			return true
+		}
+		if !canAllocPartition(node) {
+			return true
 		}
+		nodes = append(nodes, node)
 		return true
 	})
+
+	if len(nodes) < replicaNum {
+		err = fmt.Errorf("action[%vNodeSelector::Select] no enough writable hosts,replicaNum:%v  MatchNodeCount:%v  ",
+			s.GetName(), replicaNum, len(nodes))
+		return
+	}
+
+	distinctIpSet := make(map[string]struct{})
 	orderHosts := make([]string, 0)
 	for len(orderHosts) < replicaNum {
-		if len(nodes)+len(orderHosts) < replicaNum {
-			break
-		}
-		index, node := s.selectOneNode(nodes)
-		if index != 0 {
-			nodes[0], nodes[index] = node, nodes[0]
-		}
-		nodes = nodes[1:]
-		if !canAllocPartition(node) {
-			continue
+		for {
+			index, node := s.selectOneNode(nodes, distinctIpSet)
+			if index == -1 {
+				break
+			}
+
+			if index != 0 {
+				nodes[0], nodes[index] = node, nodes[0]
+			}
+			nodes = nodes[1:]
+
+			orderHosts = append(orderHosts, node.GetAddr())
+			node.SelectNodeForWrite()
+			peer := proto.Peer{ID: node.GetID(), Addr: node.GetAddr(), ReplicaPort: node.GetReplicaPort(), HeartbeatPort: node.GetHeartbeatPort()}
+			peers = append(peers, peer)
+			if len(orderHosts) == replicaNum {
+				break
+			}
 		}
-		orderHosts = append(orderHosts, node.GetAddr())
-		node.SelectNodeForWrite()
-		peer := proto.Peer{ID: node.GetID(), Addr: node.GetAddr(), ReplicaPort: node.GetReplicaPort(), HeartbeatPort: node.GetHeartbeatPort()}
-		peers = append(peers, peer)
+		distinctIpSet = make(map[string]struct{})
```

Comment:
The distInctIpSet seems reset after the inner loop everytime, so there's a chance to select two nodes with the same ip address?